lightning/restore/restore.go

```diff
@@ -304,48 +305,209 @@ outside:
+}
 
-			tablesSchema := make(map[string]string)
+func (worker *restoreSchemaWorker) makeJobs(dbMetas []*mydump.MDDatabaseMeta) {
+	//TODO: maybe we should put these create statements into a transaction
+	select {
+	case <-worker.ctx.Done():
+		return
+	default:
+		// 1. restore databases, execute statements concurrency
+		for _, dbMeta := range dbMetas {
+			restoreSchemaJob := &schemaJob{
+				dbName:  dbMeta.Name,
+				session: worker.getSession(dbMeta.Name),
+				stmts:   make([]*schemaStmt, 0, 1),
+			}
+			restoreSchemaJob.stmts = append(restoreSchemaJob.stmts, &schemaStmt{
+				stmtType: schemaCreateDatabase,
+				sql:      createDatabaseIfNotExistStmt(dbMeta.Name),
+			})
+			worker.wg.Add(1)
+			worker.jobCh <- restoreSchemaJob
+		}
+		worker.wg.Wait()
+		// 2. restore tables, execute statements concurrency
+		for _, dbMeta := range dbMetas {
 			for _, tblMeta := range dbMeta.Tables {
-				tablesSchema[tblMeta.Name] = tblMeta.GetSchema(ctx, rc.store)
+				sql := tblMeta.GetSchema(worker.ctx, worker.store)
+				if sql != "" {
+					stmts, err := createTableIfNotExistsStmt(worker.glue.GetParser(), sql, dbMeta.Name, tblMeta.Name)
+					if err != nil {
+						worker.throw(err)
+					}
+					for _, sql := range stmts {
+						restoreSchemaJob := &schemaJob{
+							dbName:  dbMeta.Name,
+							session: worker.getSession(dbMeta.Name),
+							stmts:   make([]*schemaStmt, 0, 1),
+						}
+						restoreSchemaJob.stmts = append(restoreSchemaJob.stmts, &schemaStmt{
+							tblName:  tblMeta.Name,
+							stmtType: schemaCreateTable,
+							sql:      sql,
+						})
+						worker.wg.Add(1)
+						worker.jobCh <- restoreSchemaJob
+					}
+				}
 			}
-			err := InitSchema(ctx, rc.tidbGlue, dbMeta.Name, tablesSchema)
-
-			task.End(zap.ErrorLevel, err)
-			if err != nil {
-				return errors.Annotatef(err, "restore table schema %s failed", dbMeta.Name)
+		}
+		worker.wg.Wait()
+		// 3. restore views. Since views can cross database we must restore views after all table schemas are restored.
+		for _, dbMeta := range dbMetas {
+			restoreSchemaJob := &schemaJob{
+				dbName:  dbMeta.Name,
+				session: worker.getSession(dbMeta.Name),
+				stmts:   make([]*schemaStmt, 0), // sadly, we can't pre-alloc precise value of cap
+			}
+			for _, viewMeta := range dbMeta.Views {
+				sql := viewMeta.GetSchema(worker.ctx, worker.store)
+				if sql != "" {
+					stmts, err := createTableIfNotExistsStmt(worker.glue.GetParser(), sql, dbMeta.Name, viewMeta.Name)
+					if err != nil {
+						worker.throw(err)
+					}
+					for _, sql := range stmts {
+						restoreSchemaJob.stmts = append(restoreSchemaJob.stmts, &schemaStmt{
+							tblName:  viewMeta.Name,
+							stmtType: schemaCreateView,
+							sql:      sql,
+						})
+					}
+				}
 			}
+			worker.wg.Add(1)
+			worker.jobCh <- restoreSchemaJob
+			// we don't support restore views concurrency, cauz it maybe will raise a error
+			worker.wg.Wait()
 		}
+		// for quit
+		worker.throw(nil)
```

Comment:
could we `worker.quit`?

Reply:
Don't do that, whole logic of quit in `restoreSchemaWorker.wait()`, we shouldn't just call `worker.quit()`.

Reply:
`worker.quit()` is the cancel function of context, it also could trigger `restoreSchemaWorker.wait()`

Reply:
Yes, it's true. If you prefer that I can move `worker.quit()` to select err case, I will replace `worker.throw(nil)` to `worker.quit()`